pkg/kubelet/cm/dra/manager.go

```diff
@@ -175,46 +266,53 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 				return fmt.Errorf("NodePrepareResources failed for claim %s/%s: %s", reqClaim.Namespace, reqClaim.Name, result.Error)
 			}
 
-			claimInfo := claimInfos[types.UID(claimUID)]
+			claim := resourceClaims[types.UID(claimUID)]
 
-			// Add the CDI Devices returned by NodePrepareResources to
-			// the claimInfo object.
-			err = claimInfo.addCDIDevices(pluginName, result.GetCDIDevices())
+			// Add the prepared CDI devices to the claim info
+			err := m.cache.withLock(func() error {
+				info, exists := m.cache.get(claim.Name, claim.Namespace)
+				if !exists {
+					return fmt.Errorf("unable to get claim info for claim %s in namespace %s", claim.Name, claim.Namespace)
+				}
+				if err := info.setCDIDevices(pluginName, result.GetCDIDevices()); err != nil {
+					return fmt.Errorf("unable to add CDI devices for plugin %s of claim %s in namespace %s", pluginName, claim.Name, claim.Namespace)
+				}
+				return nil
+			})
 			if err != nil {
-				return fmt.Errorf("failed to add CDIDevices to claimInfo %+v: %+v", claimInfo, err)
+				return fmt.Errorf("locked cache operation: %w", err)
 			}
-			// mark claim as (successfully) prepared by manager, so next time we don't prepare it.
-			claimInfo.prepared = true
-
-			// TODO: We (re)add the claimInfo object to the cache and
-			// sync it to the checkpoint *after* the
-			// NodePrepareResources call has completed. This will cause
-			// issues if the kubelet gets restarted between
-			// NodePrepareResources and syncToCheckpoint. It will result
-			// in not calling NodeUnprepareResources for this claim
-			// because no claimInfo will be synced back to the cache
-			// for it after the restart. We need to resolve this issue
-			// before moving to beta.
-			m.cache.add(claimInfo)
-		}
-
-		// Checkpoint to reduce redundant calls to
-		// NodePrepareResources after a kubelet restart.
-		err = m.cache.syncToCheckpoint()
-		if err != nil {
-			return fmt.Errorf("failed to checkpoint claimInfo state, err: %+v", err)
 		}
 
 		unfinished := len(claims) - len(response.Claims)
 		if unfinished != 0 {
 			return fmt.Errorf("NodePrepareResources left out %d claims", unfinished)
 		}
 	}
-	// Checkpoint to capture all of the previous addPodReference() calls.
-	err := m.cache.syncToCheckpoint()
+
+	// Atomically perform some operations on the claimInfo cache.
+	err := m.cache.withLock(func() error {
+		// Mark all pod claims as prepared.
+		for _, claim := range resourceClaims {
+			info, exists := m.cache.get(claim.Name, claim.Namespace)
+			if !exists {
+				return fmt.Errorf("unable to get claim info for claim %s in namespace %s", claim.Name, claim.Namespace)
+			}
+			info.setPrepared()
```

Comment:
@klueska We mark claiminfo as prepared only if `NodePrepareResources` succeeded for all plugins. I think we incorrectly handle the case of partial preparation - when the `NodePrepareResources` succeeded only for some plugins. If pod is deleted, Kubelet will not call `NodeUnprepareResources` for those plugins as this information is not cached.

Reply:
The `prepared` field is only used to avoid erroneous calls to `PrepareResources` -- `UnprepareResources` doesn't look at it at all. If we have ever created a claimInfo object and added a pod reference to it, we _will_ get an `UnprepareResources` call for it (either in sync with the deletion or in the reconcile loop).

Reply:
It makes sense to me now. Thanks for the explanations.